Mathlib/Topology/Constructible.lean

```diff
@@ -0,0 +1,303 @@
+  · rw [IsEmbedding.subtypeVal.isCompact_iff, image_preimage_eq_inter_range,
+      Subtype.range_coe_subtype, setOf_mem_eq, inter_comm]
+    exact hs htcomp htopen
+  · simpa using (hs.isCompact_preimage_of_isOpen htopen htcomp).image continuous_subtype_val
+
+@[stacks 005B]
+lemma IsRetroCompact.image_of_isEmbedding (hs : IsRetroCompact s) (hfemb : IsEmbedding f)
+    (hfcomp : IsRetroCompact (range f)) : IsRetroCompact (f '' s) := by
+  rintro U hUcomp hUopen
+  rw [← image_inter_preimage, ← hfemb.isCompact_iff]
+  refine hs ?_ <| hUopen.preimage hfemb.continuous
+  rw [hfemb.isCompact_iff, image_preimage_eq_inter_range, inter_comm]
+  exact hfcomp hUcomp hUopen
+
+@[stacks 005J] -- Extracted from the proof of `005J`
+lemma IsRetroCompact.preimage_of_isOpenEmbedding {s : Set β} (hf : IsOpenEmbedding f)
+    (hs : IsRetroCompact s) : IsRetroCompact (f ⁻¹' s) := by
+  rintro U hUcomp hUopen
+  rw [hf.isCompact_iff, image_preimage_inter]
+  exact hs (hUcomp.image hf.continuous) <| hf.isOpenMap _ hUopen
+
+@[stacks 09YE] -- Extracted from the proof of `09YE`
+lemma IsRetroCompact.preimage_of_isClosedEmbedding {s : Set β} (hf : IsClosedEmbedding f)
+    (hf' : IsCompact (range f)ᶜ) (hs : IsRetroCompact s)  : IsRetroCompact (f ⁻¹' s) := by
+  rintro U hUcomp hUopen
+  have hfUopen : IsOpen (f '' U ∪ (range f)ᶜ) := by
+    simpa [← range_diff_image hf.injective, sdiff_eq, compl_inter, union_comm]
+      using (hf.isClosedMap _ hUopen.isClosed_compl).isOpen_compl
+  have hfUcomp : IsCompact (f '' U ∪ (range f)ᶜ) := (hUcomp.image hf.continuous).union hf'
+  simpa [inter_union_distrib_left, inter_left_comm, inter_eq_right.2 (image_subset_range ..),
+    hf.isCompact_iff, image_preimage_inter] using (hs hfUcomp hfUopen).inter_left hf.isClosed_range
+
+/-! ### Constructible sets -/
+
+/-- A constructible set is a set that can be written as the
+finite union/finite intersection/complement of open retrocompact sets.
+
+In other words, constructible sets form the boolean subalgebra generated by open retrocompact sets.
+-/
+def IsConstructible (s : Set α) : Prop :=
+  s ∈ BooleanSubalgebra.closure {U | IsOpen U ∧ IsRetroCompact U}
+
+@[simp]
+protected lemma IsConstructible.empty : IsConstructible (∅ : Set α) := BooleanSubalgebra.bot_mem
+
+@[simp]
+protected lemma IsConstructible.univ : IsConstructible (univ : Set α) := BooleanSubalgebra.top_mem
+
+lemma IsConstructible.union : IsConstructible s → IsConstructible t → IsConstructible (s ∪ t) :=
+  BooleanSubalgebra.sup_mem
+
+lemma IsConstructible.inter : IsConstructible s → IsConstructible t → IsConstructible (s ∩ t) :=
+  BooleanSubalgebra.inf_mem
+
+lemma IsConstructible.sdiff : IsConstructible s → IsConstructible t → IsConstructible (s \ t) :=
+  BooleanSubalgebra.sdiff_mem
+
+lemma IsConstructible.himp : IsConstructible s → IsConstructible t → IsConstructible (s ⇨ t) :=
+  BooleanSubalgebra.himp_mem
+
+@[simp] lemma isConstructible_compl : IsConstructible sᶜ ↔ IsConstructible s :=
```

Comment:
shouldn't it be `isConstructible_compl_iff`?

Reply:
`_iff` and `_eq` are only necessary when ambiguity would otherwise arise